fixes some typos and types

xeno-by opened the pull request:
review @retronym /cc @densh

xeno-by committed:
quasiquotes no longer evaluate debug logs when debug logging is off

xeno-by committed:
fixes a typo in Types.scala

xeno-by committed:
introduces failsafe against endless type printing

The parent commit works around a particular problem that led to a compiler
freeze in SI-8158, whereas this commit introduces a general solution -
a cache that tracks all types that we've recursed into during printing.

I can't immediately come up with an example of a type that would be caught
by this safety net, but unknown unknowns are the worst of them all, so why not
guard against them while we can.

xeno-by committed:
SI-8158 compiler hangs printing out fancy types

Apparently, even though the compiler has safeguards against infinite type
printouts, having a depth counter, we didn’t account for the cases when
printouts are both self-referential and self-multiplying.

For one, SI-8158 provides an example of such a type, which is a structural
type that refers to itself twice in return types of its methods. At first,
printing such a type would go deeper and deeper, but then it will hit
the depth limit and start multiply indefinitely.

This commit fixes this particular problem by recognizing self-references
as this.type’s and printing them out as such. The subsequent commit will
introduce a more general facility.

xeno-by commented:
@som-snytt What kind of puns would you recommend here? :)

soc commented on src/reflect/scala/reflect/internal/tpe/TypeToStrings.scala:
```diff
@@ -14,8 +16,15 @@ private[internal] trait TypeToStrings {
   def tostringRecursions_=(value: Int) = _tostringRecursions = value
 
+  private var _tostringSubjects = HashSet[Type]()
+  def tostringSubjects = _tostringSubjects
```
`toStringSubjects`?

xeno-by replied:
Stays consistent with `tostringRecursions`

densh replied:
I'd vote to change all of them to toStringFoo rather than having more of tostringFoo.

xeno-by commented:
PLS REBUILD ALL

scala-jenkins commented:
(kitty-note-to-self: ignore 32737356)
:cat: Roger! Rebuilding pr-scala for 88c2994f, 6a6b485f, 936d60a2, 35300b46. :rotating_light:

retronym commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -2529,10 +2533,11 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
+        case PolyType(tparams, res)    => typeParamsString(tp) + infoString(res)
+        case NullaryMethodType(res)    => infoString(res)
+        case MethodType(params, res)   => valueParamsString(tp) + infoString(res)
+        case _ if isStructuralThisType => ": this.type"
```
I don't think that `this.type` is correct here.

When I typecheck the pattern of code in the test case, I get:

``` scala
scala> val x = new { def t = this }
warning: there were 1 feature warning(s); re-run with -feature for details
x: $anon forSome { type $anon <: AnyRef{def t: $anon} } = $anon$1@38b9f79a
```

Here, an existential has been introduced to capture the f-bounded-ness.

Looking at that type from the REPL:

```
scala> type bound = X forSome { type X <: { def foo: X } }
warning: there were 1 feature warning(s); re-run with -feature for details
defined type alias bound

scala> val X = typeOf[bound].dealias.typeSymbol
X: $r.intp.global.Symbol = type X

scala> X.isFBounded
res42: Boolean = false // REALLY?
```

xeno-by replied:
Well, I don't know why, but the type in question didn't feature any existentials. Method symbols of the structural type has their resultType set to `TypeRef(NoPrefix, $anon, Nil)`. Do you think `owner.name` would be more appropriate here instead of `this.type`?

retronym replied:
Yeah, I think that would be better.

xeno-by committed:
capitalizes “s” in tostring

As suggested by the reviewers, tostringXXX variables in TypeToStrings.scala
have been renamed to toStringXXX.

xeno-by committed:
addresses pull request feedback

xeno-by commented:
Addressed all the feedback

retronym commented:
I'm a little uneasy with the performance impact of the fail-safe. We end up calling `Type#toString` when we generate error messages, including, importantly, errors issued during silent mode that are never issued.

Similar code in `<:<` only turns on the detection mechanism after reaching a certain depth. Is that approach feasible here? I suspect not, given the way you described the exponential widening of the tree.

xeno-by commented:
@retronym Since we don't have time to study performance now, how about I comment the fail-safe out with a TODO?

retronym commented:
Yep, that's okay.

xeno-by committed:
temporarily disables the toStringSubjects cache

xeno-by commented:
@retronym Done

retronym commented:
LGTM

xeno-by commented:
PLS REBUILD/pr-scala@3f3014c

scala-jenkins commented:
(kitty-note-to-self: ignore 32771086)
:cat: Roger! Rebuilding pr-scala for 3f3014cf. :rotating_light:
